### PR TITLE
Issue 3279: extra set of ribbons

### LINKED ISSUE #3279: Einführung von zwei weiteren virtuellen Ribbon ("Alternative Ribbons")
Die Alternative Ribbons Funktion ermöglicht es, die physischen Ribbons für mehr als nur zwei (bis zu vier) MCs zu benutzen. Hierfür werden zwei weitere Ribbon HW-Sources eingefügt, die die beiden physikalische Ribbons mitbenutzen. Diese heißen Ribbon 3 und 4. Es gibt also insgesamt vier (virtuelle) Ribbon HW-Sources, die den zwei physikalischen Ribbons zugeordnet werden können. Über Funct-Taster an der Base-Unit kann zwischen Ribbon 1/2 und Ribbon 3/4 gewechselt werden. Hierdurch können die Ribbons vielfältiger verwendet werden und die MCs E und F werden aufgewertet. Ist einem Ribbon kein MC zugewiesen, wird das durch “not assigned” im SOLED sichtbar gemacht. Die Umschaltbarkeit zwischen absolute und relative Mode der Ribbons in der Baseunit entfällt (wird an anderer Stelle ermöglicht).

Alles weitere im angehängten PDF. 

[Alternative Ribbons 2.pdf](https://github.com/nonlinear-labs-dev/C15/files/9461680/Alternative.Ribbons.2.pdf)

### PR DESCRIPTION
**Pull request by schlenkibus**
closes #3279

### CODE REVIEW

**Comment by schlenkibus** [comment]
@MatthiasSeeber a preview build is currently in the oven!

**Comment by majjam** [comment]
i tried this build and have two points: 

**First [small bug]:**
**led indicators from a MC with  ribbon 2 will be shown at ribbon4 too and vice versa.** (no 1 to 3 correlation )
1. assign pedal1 to mcA
2. assign ribbon 2 to mcA
3. ribbon 1, 3 and 4 are not assigned anywhere
select boled view of  rib3 and rib 4
led indicators move around when pedal1 is moved

**Second [enhancement]:** 
Additional one proposal from me:
Preset should save the ribbon view position: 3and4 or 1and2 (with default at 1and2-view)
This could save the player one button press in live situations.
Now if i select 3and4 and scroll through some other presets i do see "not assigned" most of the times.

**Comment by schlenkibus** [comment]
@majjam did you use the build I provided in this thread or did you build your own update?
If you used the .tar provided by me, did you encounter any update-problems?
If so: what version where you running previously?

**Comment by majjam** [comment]
@schlenkibus  I used the provided build from here.
i did not find any update problems. 
i came from milestone-maintenance  7921e008f194931d5055e816f4494b9827425dfe (before pi4)

**Comment by schlenkibus** [comment]
> @schlenkibus  I used the provided build from here.
> i did not find any update problems. 
> i came from milestone-maintenance  7921e008f194931d5055e816f4494b9827425dfe (before pi4)
> 
> 

Thank you for the quick response!

**Review by MatthiasSeeber — CHANGES_REQUESTED** [review]
No Code Review yet, tested via IDE (PG, BBBB, Midi-Over-Ip, AudioEngine).
So far found issues:
- [x] MIDI Settings: this doesn't work reliably. I assigned Rib 3/4 to default Rib 1/2 CCs and they didn't move. After several tryouts, they finally did move when Rib 3 = Rib 1 = CC 24, Rib 4 = Rib 2 = CC 25. Then I reassigned everything to default. Still, Rib 3 and 4 are moving as well
- [ ] Crash in PG and Midi-Over-Ip (see 
[midi-over-ip-crash.txt](https://github.com/nonlinear-labs-dev/C15/files/9706860/midi-over-ip-crash.txt) and 
[playground-crash.zip](https://github.com/nonlinear-labs-dev/C15/files/9706879/playground-crash.zip) - attention, big file)
- [x] Relative Indicator (WebUI ok, Boled/Soled missing)
- [ ] unassigned Ribbons (Soled: dotted lines missing)
- [x] unassigned MCs (unfilled MC symbols, missing)
- [ ] is Ribbon 1/3 in Edit mode working? Hard to say, Soled Screens react accordingly, but does touching and changing the Ribbon control the selected parameter correctly? Via BBBB, the two long Ribbon sliders are useless so via IDE there is no way to verify?
- [x] Ribbon Menus (WebUI, are working - I find it helpful to separate Return Behavior and Absolute/Relative Mode, deviating from the proposal. Maybe, a visual separator would be helpful too?)
- [ ] MC Screen: Soft Button 3 is labeled "more" and leads to a subscreen and back. SB 4 leads to HW Amt Screen (more or less as before) and SB 3 leads back. I'm not sure how consistent this is, but it is working.

File: projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.h [hunk 1]
```diff
@@ -37,7 +37,7 @@ inline constexpr bool VOICE_FADE_INTERPOLATION = true;
 inline constexpr bool LOG_MISSING = false;
 inline constexpr bool LOG_FAIL = false;
 inline constexpr bool LOG_INIT = false;
-inline constexpr bool LOG_MIDI_TCD = false;
+inline constexpr bool LOG_MIDI_TCD = true;
```
Comment (MatthiasSeeber) [inline projects/epc/audio-engine/src/synth/c15-audio-engine/dsp_host_dual.h:40]: Cleanup: don't forget to deactivate this before merge.

File: projects/epc/audio-engine/src/synth/input/InputEventStage.h [hunk 2]
```diff
@@ -14,6 +14,8 @@ class MidiRuntimeOptions;
 
 class InputEventStage
 {
+ private:
+  constexpr static auto NUM_HW = 10;
```
Comment (MatthiasSeeber) [inline projects/epc/audio-engine/src/synth/input/InputEventStage.h:18]: :+1:  As mentioned, could be provided by configuration.

**Comment by schlenkibus** [comment]
- [ ] select modulateable parameter after selecting unmodulateable parameter -> the parameter name is not highlighted

**Review by MatthiasSeeber — COMMENTED** [review]
So far found:
- :ok: unassigned Ribbons --> dotted Lines in Soled
- Boled, MC Screen: press SB4 twice and the upper right Group Label will indicate changes to the Macro Group (`!Macro Control`), although only the focus was changed?
- the changes to the HW Amounts and removal of several corresponding Screens means that there is no more Parameter Screen for HW Amounts. Consequence: Parameter Info for HW Amounts can only be reached via WebUI (if in HW Amt Mode, the MC Screen could display HW Amount Info instead of MC info)

MIDI Settings
- [ ] still buggy:
assign Ribbon1 to some CC and send something --> Ribbon1 will change, but also Ribbon3 (and even Ribbon4? Similarly, do this with Ribbon2 --> Ribbon4 will change - without any MC Assignments, even if Ribbon 3/4 have no CC set)

unassigned MC --> unfilled Symbol
- [x] MC Boled
- [x] Soled for assigned Ribbon
- [ ] HW Amounts (like "Pedal 1 to MC A") also use filled symbols. Does this apply here as well?

**Comment by schlenkibus** [comment]
Bezueglich der Filled MC-Symbole in den Routings bin ich ueberfragt, @nonlinear-labs was ist deine Meinung?

MIDI:
Test auf meinem Device: 
Ribbon 1: CC18
Ribbon 2: CC25
Ribbon 3: CC26
Ribbon 4: CC27

Ich habe den Ribbon 1 ueber die HW bewegt und auf meinen PC einen Midi-Clip aufgenommen.
Wenn ich den Clip replaye bewegt sich ribbon 1 wie erwartet. Auch das umschalten auf Ribbon 3/4 fuehrt zu den gewuenschten Ergebnissen: Ribbon 3 und 4 bewegen sich nicht.

**Comment by MatthiasSeeber** [comment]
Hier, die vier Pedale sind genauso assigned wie die vier Ribbons. Ich gehe die CCs 20 ... 23 nacheinander durch. Ich habe diesmal midi-over-ip nicht mit dabei, vorher lief das bei mir immer mit..

[Screencast from 05.10.2022 14:22:49.webm](https://user-images.githubusercontent.com/37212609/194059498-6f5480e0-3e46-437a-a73b-6199fc414845.webm)

**Comment by davidm2894** [comment]
- changing mode of one ribbon to relative has impact on both ribbons of this ribbon pair. I assume that the mode of the physical ribbon not the virtual ribbon is changed. when changed back to absolute mode, the LED ribbons of the second ribbon show the position of the first ribbon (if second ribbon of a pair is selected)
- the superscripted "r" is quite small and for the readability it would be better to leave the pixel in top left corner blank. Is this a symbol from the font?
- if you change the MC parameter of a MC connected to one of the ribbons of the second pair (2/4), both ribbons move.
- CC XX misses for ribbons 3/4 in HW Source Screen
- the HW Sel field in MC screen shows the superscripted "r", it would be better if it's not shown in this place

File: projects/epc/audio-engine/src/synth/input/InputEventStage.cpp [hunk 3]
```diff
@@ -785,9 +798,22 @@ HardwareSource InputEventStage::parameterIDToHWID(int id)
...
-  const auto routingIndex = static_cast<RoutingIndex>(hwID);
+  if constexpr (LOG_INPUT_EVENT_STAGE)
+  {
+    nltools::Log::error(__PRETTY_FUNCTION__, "hwid:", toString(hwID), "pos:", pos, "source:", toString(source), "hwid idx:", static_cast<unsigned int>(hwID));
+  }
+
+  RoutingIndex routingIndex = RoutingIndex::LENGTH;
 
-  auto sendToDSP = [&](auto source, auto hwID, auto wasPrim, auto wasSplit) {
+  if(hwID == HardwareSource::RIBBON3)
```
Comment (hhoegelo) [inline projects/epc/audio-engine/src/synth/input/InputEventStage.cpp:808]: consider introducing a translateXtoY method

File: projects/epc/audio-engine/src/synth/input/TCDDecoder.cpp [hunk 4]
```diff
@@ -25,21 +25,35 @@ bool TCDDecoder::decode(const MidiEvent &event)
...
+      }
+
       const auto behaviour = m_dsp->getBehaviour(hwid);
 
       if(behaviour == C15::Properties::HW_Return_Behavior::Center)
       {
         value = (2.0f * value) - 1.0f;
       }
 
+      keyOrController = channel;
```
Comment (hhoegelo) [inline projects/epc/audio-engine/src/synth/input/TCDDecoder.cpp:48]: consider m_ for members

Comment (hhoegelo) [reply projects/epc/audio-engine/src/synth/input/TCDDecoder.cpp:48]: keyOrController should be called m_key and not be assigned if tcd message is not a key (maybe make it an optional or have a special INVALID value?)

File: projects/epc/playground/src/use-cases/EditBufferUseCases.cpp [hunk 5]
```diff
@@ -409,3 +409,9 @@ void EditBufferUseCases::setModulationSourceOfAll(MacroControls controls)
     }
   }
 }
+
+void EditBufferUseCases::selectLastSelectedMacroControlParameter()
```
Comment (hhoegelo) [inline projects/epc/playground/src/use-cases/EditBufferUseCases.cpp:413]: unused?

File: projects/epc/playground/src/presets/EditBuffer.cpp [hunk 6]
```diff
@@ -1756,3 +1763,8 @@ ParameterDB &EditBuffer::getParameterDB()
 {
   return m_parameterDB;
 }
+
+ParameterId EditBuffer::getLastSelectedMacroId() const
```
Comment (hhoegelo) [inline projects/epc/playground/src/presets/EditBuffer.cpp:1767]: unused?

File: projects/epc/playground/src/presets/EditBuffer.cpp [hunk 7]
```diff
@@ -50,6 +51,7 @@ EditBuffer::EditBuffer(PresetManager *parent, Settings &settings, std::unique_pt
     , m_loadedPresetLog(std::make_unique<LoadedPresetLog>())
     , m_settings { settings }
     , m_audioEngineProxyContainer { aeContainer }
+    , m_lastSelectedMacroControl { C15::PID::MC_A, VoiceGroup::Global }
```
Comment (hhoegelo) [inline projects/epc/playground/src/presets/EditBuffer.cpp:54]: unused?

File: projects/epc/playground/src/proxies/hwui/base-unit/LowerRibbon.cpp [hunk 8]
```diff
@@ -12,28 +14,68 @@
 
 static Parameter *getParameter()
 {
-  return Application::get().getPresetManager()->getEditBuffer()->findParameterByID(
-      HardwareSourcesGroup::getLowerRibbonParameterID());
+  if(Application::get().getSettings()->getSetting<SelectedRibbonsSetting>()->get() == SelectedRibbons::Ribbon1_2)
+  {
+    return Application::get().getPresetManager()->getEditBuffer()->findParameterByID(
+        HardwareSourcesGroup::getLowerRibbonParameterID());
```
Comment (hhoegelo) [inline projects/epc/playground/src/proxies/hwui/base-unit/LowerRibbon.cpp:20]: consider being consistent with getLowerRibbon4ParameterID

File: projects/epc/playground/src/proxies/hwui/base-unit/UpperRibbon.cpp [hunk 9]
```diff
@@ -61,8 +79,17 @@ Parameter* UpperRibbon::getResponsibleParameter()
     return Application::get().getPresetManager()->getEditBuffer()->getSelected(
         Application::get().getVGManager()->getCurrentVoiceGroup());
 
-  return Application::get().getPresetManager()->getEditBuffer()->findParameterByID(
-      HardwareSourcesGroup::getUpperRibbonParameterID());
+  auto selected = Application::get().getSettings()->getSetting<SelectedRibbonsSetting>();
+  if(selected->get() == SelectedRibbons::Ribbon1_2)
+  {
+    return Application::get().getPresetManager()->getEditBuffer()->findParameterByID(
+        HardwareSourcesGroup::getUpperRibbonParameterID());
```
Comment (hhoegelo) [inline projects/epc/playground/src/proxies/hwui/base-unit/UpperRibbon.cpp:86]: consistent naming

File: projects/epc/playground/src/proxies/hwui/base-unit/soled/SelectedRibbonsIndication.cpp [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+void SelectedRibbonsIndication::onSettingChanged(const Setting* s)
+{
+  if(auto setting = dynamic_cast<const SelectedRibbonsSetting*>(s))
+  {
+    m_isRibbon1_2_selected = setting->get() == SelectedRibbons::Ribbon1_2;
+  }
+}
+
+void drawDottedLine(FrameBuffer& fb, int l, int t, int lenght)
```
Comment (hhoegelo) [inline projects/epc/playground/src/proxies/hwui/base-unit/soled/SelectedRibbonsIndication.cpp:29]: typo length

File: projects/epc/playground/src/proxies/hwui/base-unit/soled/SelectedRibbonsIndication.cpp [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+  {
+    drawDottedLine(fb, l, t, length);
+  }
+}
+
+bool isRibbonAssigned(int pid)
+{
+  auto ribbon = Application::get().getPresetManager()->getEditBuffer()->findAndCastParameterByID<RibbonParameter>(
+      { pid, VoiceGroup::Global });
+  return ribbon->getCurrentModulatingMacroControlString() != "not assigned";
```
Comment (hhoegelo) [inline projects/epc/playground/src/proxies/hwui/base-unit/soled/SelectedRibbonsIndication.cpp:51]: this smells

File: projects/epc/playground/src/proxies/hwui/descriptive-layouts/LayoutFactory.cpp [hunk 12]
```diff
@@ -68,8 +68,8 @@ namespace DescriptiveLayouts
     return *it;
   }
 
-  std::shared_ptr<Layout> BoledLayoutFactory::instantiate(FocusAndMode fam)
+  Layout* BoledLayoutFactory::instantiate(FocusAndMode fam)
```
Comment (hhoegelo) [inline projects/epc/playground/src/proxies/hwui/descriptive-layouts/LayoutFactory.cpp:71]: investigate: whats motivation to return raw pointer here?

File: projects/epc/playground/src/proxies/hwui/panel-unit/boled/parameter-screens/MacroControlParameterLayouts.cpp [hunk 13]
```diff
@@ -61,6 +76,27 @@ void MacroControlParameterLayout2::copyFrom(Layout *other)
   if(auto p = dynamic_cast<MacroControlParameterLayout2 *>(other))
     setMode(p->m_mode);
 
+  if(auto p = dynamic_cast<ModulationRouterParameterSelectLayout2*>(other))
+  {
+    nltools::Log::error(toString(p->getMode()));
```
Comment (hhoegelo) [inline projects/epc/playground/src/proxies/hwui/panel-unit/boled/parameter-screens/MacroControlParameterLayouts.cpp:81]: is that really an error?

File: projects/epc/playground/src/proxies/hwui/panel-unit/boled/parameter-screens/MacroControlParameterLayouts.h [hunk 14]
```diff
@@ -7,26 +7,33 @@ class MacroControlParameterLayout2 : public virtual ParameterLayout2
...
   bool onRotary(int inc, ButtonModifiers modifiers) override;
   Parameter *getCurrentEditParameter() const override;
 
-  enum class Mode
-  {
-    MacroControlValue,
-    PlayControlPosition,
-    PlayControlSelection,
-    PlayControlAmount
+  struct InitialParameterSelection {
```
Comment (hhoegelo) [inline projects/epc/playground/src/proxies/hwui/panel-unit/boled/parameter-screens/MacroControlParameterLayouts.h:29]: huh?

File: projects/epc/playground/src/proxies/hwui/panel-unit/boled/parameter-screens/MacroControlParameterLayouts.h [hunk 15]
```diff
@@ -35,6 +42,8 @@ class MacroControlParameterLayout2 : public virtual ParameterLayout2
   void setButtonA(Button *button);
   virtual Control *createMCAssignmentIndicator();
 
+  std::optional<InitialParameterSelection> m_initialParameterSelection = std::nullopt;
```
Comment (hhoegelo) [inline projects/epc/playground/src/proxies/hwui/panel-unit/boled/parameter-screens/MacroControlParameterLayouts.h:45]: huh?

File: projects/epc/playground/src/proxies/hwui/panel-unit/boled/parameter-screens/controls/HWSourceAmountCarousel.cpp [hunk 16]
```diff
@@ -21,61 +22,86 @@ HWSourceAmountCarousel::~HWSourceAmountCarousel()
...
 void HWSourceAmountCarousel::antiTurn()
 {
-  auto vg = Application::get().getVGManager()->getCurrentVoiceGroup();
-  auto eb = Application::get().getPresetManager()->getEditBuffer();
-  if(auto p = dynamic_cast<MacroControlParameter *>(eb->getSelected(vg)))
-    p->toggleUiSelectedHardwareSource(-1);
+  if(isSelectedHighlighted())
+    doTurn(-1);
+  else
+    doTurn(0);
```
Comment (hhoegelo) [inline projects/epc/playground/src/proxies/hwui/panel-unit/boled/parameter-screens/controls/HWSourceAmountCarousel.cpp:55]: pls check if doTurn(0) actually does something (also in turn())

File: projects/epc/playground/src/proxies/playcontroller/PlaycontrollerProxy.cpp [hunk 17]
```diff
@@ -224,8 +225,10 @@ Parameter *PlaycontrollerProxy::findPhysicalControlParameterFromPlaycontrollerHW
         return HardwareSourcesGroup::getUpperRibbonParameterID();
       case HW_SOURCE_ID_RIBBON_2:
         return HardwareSourcesGroup::getLowerRibbonParameterID();
-      case HW_SOURCE_ID_PEDAL_5:
-      case HW_SOURCE_ID_PEDAL_6:
+      case 8:  //meh
```
Comment (hhoegelo) [inline projects/epc/playground/src/proxies/playcontroller/PlaycontrollerProxy.cpp:228]: Pls try to not use Playcontroller Domain enum in Audioengineproxy context

**Comment by davidm2894** [comment]
I had a crash with the Preview Build from this PR. Error Journal is attached and was downloaded just right after the crash. Banks were reloaded and sound was gone for a moment. 

[C15-journal.tar.gz.txt](https://github.com/nonlinear-labs-dev/C15/files/9732729/C15-journal.tar.gz.txt)

File: projects/epc/playground/src/use-cases/EditBufferUseCases.cpp (hunk 5 above)
Comment (schlenkibus) [reply projects/epc/playground/src/use-cases/EditBufferUseCases.cpp:413]: still used!

**Review by hhoegelo — APPROVED** [review]

**Pull request merged**